File: lab3/part1/tiger.c
#include "animal.h"
#include <stdlib.h>

char const* name(struct Animal* this) {
    return this->name;
}

char const* greet(void) {
    return "growl!";
}

char const* menu(void) {
    return "warm milk";
}

PTRFUN table[] = {
    name,
    greet,
    menu
};

void construct(struct Animal* memory, const char* name) {
    memory->name = name;
    memory->vtable = table;
}

struct Animal* create(char const* name) {
    struct Animal* tiger = (struct Animal*) malloc(sizeof(struct Animal));
    construct(tiger, name);
    return tiger;
}

File: lab3/part1/task4/plugins/Tiger.java
package task4.plugins;

import task4.Animal;

public final class Tiger extends Animal {

    private final String name;

    public Tiger(String name) {
        this.name = name;
    }

    @Override
    public String name() {
        return this.name;
    }

    @Override
    public String greet() {
        return "growl!";
    }

    @Override
    public String menu() {
        return "warm milk";
    }
}
